File: java/src/main/java/com/example/design_pattern/Proxy.java
package com.example.design_pattern;


import java.util.*;

public class Proxy {
	
	private static class Proxy_ {

		private static interface ServiceInterface {
			void operation();
		}

		private static class Service implements ServiceInterface {
			public void operation(){
			}
		}

		private static class ServiceProxy implements ServiceInterface {
			
			ServiceInterface service;

			ServiceProxy(ServiceInterface service){
				this.service = service;
			}

			private boolean checkAccess(){
				System.out.println("Proxy : checking access prior to firing real request.");
				return true;
			}

			private void logAccess(){
				System.out.println("Proxy : Logging the time of request");
			}

			public void operation(){
				if(checkAccess()){
					service.operation();
					logAccess();
				}
			}
		}

		public static void main(){
			ServiceInterface proxy = new ServiceProxy(new Service());
			proxy.operation();
		}		
	}

	private static class AttributeProxy{

		private static class Property <T> {
			T value;

			Property(T value){
				this.value = value;
			}

			T get(){
				System.out.println("get : something to do");
				return value;
			}

			void set(T value){
				System.out.println("set : something to do");
				this.value = value;
			}
		}

		private static class Creature {
			private Property<Integer> strength;
			private Property<Integer> agility;

			Creature(int strength, int agility){
				this.strength = new Property<Integer>(strength);
				this.agility = new Property<Integer>(agility);
			}

			public void setStrength(int strength){
				this.strength.set(strength);
			}

			public void setAgility(int agility){
				this.agility.set(agility);
			}

			public int getStrength(){
				return strength.get();
			}

			public int getAgility(){
				return agility.get();
			}
		}

		public static void main(){
			Creature creature = new Creature(10,20);
			int str = creature.getStrength();
			creature.setAgility(10);
		}
	}

	private static class CachedProxy {

		private class VideoInfo{}
		private class VideoData{}

		private interface YoutubeService {
			List<String> getVideos();
			VideoInfo getVideoInfo(int id);
			void download(int id);
		}

		private class YoutubeServiceImpl implements YoutubeService{
			public List<String> getVideos(){
				return List.of();
			}

			public VideoInfo getVideoInfo(int id){
				return new VideoInfo();
			}

			public void download(int id){
			}
		}

		private class YoutubeServiceProxy implements YoutubeService{
			
			YoutubeService service;

			private List<String> ids;
			private Map<Integer,VideoInfo> infos;
			private List<Integer> downloaded;
			private boolean needReset;

			YoutubeServiceProxy(YoutubeService service){
				this.service = service;
				infos = new HashMap<>();
				downloaded = new LinkedList<>();
				needReset = false;
			}

			public List<String> getVideos(){
				if(ids == null || needReset){
					return ids= service.getVideos();
				}
				return ids;
			}

			public VideoInfo getVideoInfo(int id){
				if(!infos.containsKey(id) || needReset){
					VideoInfo info = service.getVideoInfo(id);
					infos.put(id, info);
				}
				return infos.get(id);
			}

			public void download(int id){
				if(!downloaded.contains(id) || needReset){
					service.download(id);
					downloaded.add(id);
				}
			}
		}

		public static void main(){

		}
	}

	private static class VirtualProxy {

		private static interface Image {
			void draw();
		}

		private static class Bitmap implements Image {
			Bitmap(String filename){
				System.out.println("Loading image from "+filename);
			}

			public void draw(){
				System.out.println("Drawing image");
			}
		}

		private static class LazyBitmap implements Image {
			
			private String filename;
			private Bitmap bitmap;

			LazyBitmap(String filename){
				this.filename = filename;
				System.out.println("Lazy Bitmap from "+ filename);
			}

			public void draw(){
				System.out.println("About to draw the image");
				if(bitmap == null){
					bitmap = new Bitmap(filename);
				}
				bitmap.draw();
				System.out.println("Complete drawing the image");
			}
		}

		public static void main(){
			Image bitmap = new LazyBitmap("pokemon.png");
			bitmap.draw();
		}
	}

	public static void main(String... args){
		Proxy_.main();
		AttributeProxy.main();
		CachedProxy.main();
		VirtualProxy.main();
	}
}

File: c++/design_pattern/Proxy.cpp
#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace std;

namespace Proxy {

struct ServiceInterface {
  virtual void operation() = 0;
};

struct Service : ServiceInterface {
  void operation() override {}
};

struct ServiceProxy : ServiceInterface {
  ServiceProxy(ServiceInterface* service) : service(service) {}

  bool check_access() {
    cout << "Proxy : Checking access priror to firing real request" << endl;
    return true;
  }

  void log_access() { cout << "Proxy : Loggint the time of request" << endl; }

  void operation() override {
    if (this->check_access()) {
      this->service->operation();
      this->log_access();
    }
  }

 private:
  ServiceInterface* service;
};

void main() {
  Service service;
  ServiceProxy proxy(&service);
  proxy.operation();
}
};  // namespace Proxy

namespace AttributeProxy {

template <typename T>
struct Property {
  T value;
  Property(const T initial_value) : value(initial_value) {}

  operator T() {
    cout << "some get method doing here" << endl;
    return value;
  }

  T operator=(T new_value) {
    cout << "some set method doing here" << endl;
    return value = new_value;
  }
};

struct Creature {
  Property<int> strength{10};
  Property<int> agility{5};
};

void main() {
  Creature creature;
  creature.agility = 20;
  auto x = creature.strength;
}
};  // namespace AttributeProxy

namespace CachedProxy {

struct VideoInfo {};
struct VideoData {};
struct Video {
  int id;
  VideoInfo info;
  VideoData data;
};

struct ThirdPartyYoutubeLib {
  virtual vector<string> list_videos() = 0;
  virtual VideoInfo get_video_info(int id) = 0;
  virtual void downlaod_video(int id) = 0;
};

struct ThidPartyYoutubeClass : ThirdPartyYoutubeLib {
  vector<string> list_videos() override {
    vector<string> ret;
    return ret;
  };
  VideoInfo get_video_info(int id) override {
    VideoInfo ret;
    return ret;
  };
  void downlaod_video(int id) override {

  };
};

struct CachedYoutubeClass : ThirdPartyYoutubeLib {
  CachedYoutubeClass(ThirdPartyYoutubeLib* service) : service(service) {}

  vector<string> list_videos() override {
    if (!list_cache.size() || need_reset) {
      list_cache = service->list_videos();
    }
    return list_cache;
  };
  VideoInfo get_video_info(int id) override {
    if (!info_cache.size() || need_reset) {
      info_cache[id] = service->get_video_info(id);
    }
    return info_cache[id];
  };
  void downlaod_video(int id) override {
    if (data_cache.find(id) != end(data_cache) || need_reset) {
      service->downlaod_video(id);
    }
  };

 private:
  ThirdPartyYoutubeLib* service;
  vector<string> list_cache;
  map<int, VideoInfo> info_cache;
  map<int, VideoData> data_cache;
  bool need_reset;
};

void main() {}
};  // namespace CachedProxy

namespace VirtualProxy {

struct Image {
  virtual void draw() = 0;
};

// 생성자 호출시에 이미지를 메모리에 적재
struct Bitmap : Image {
  Bitmap(const string& filename) {
    cout << "Loading image from " << filename << endl;
  }

  void draw() override { cout << "Drawing image " << endl; }
};

// 생성자 호출에는 파일이름만 저장해 놓았다가, draw()메소드가 호출 될 떄만
// 메모리에 적재.
struct LazyBitmap : Image {
  LazyBitmap(const string& filename) : filename(filename) {
    cout << "Loading image from " << filename << endl;
  }
  ~LazyBitmap() { delete bitmap; }

  void draw() override {
    if (!bitmap) bitmap = new Bitmap(filename);
    bitmap->draw();
  }

 private:
  Bitmap* bitmap{nullptr};
  string filename;
};

void draw_image(Image& img) {
  cout << "About to draw the image" << endl;
  img.draw();
  cout << "Done drawing the image" << endl;
}

void main() {
  LazyBitmap img{"pokemon.png"};
  draw_image(img);
}
};  // namespace VirtualProxy

namespace CommunicationProxy {
struct Pingable {
  virtual wstring ping(const wstring& message) = 0;
};

struct Pong : Pingable {
  wstring ping(const wstring& message) override { return message + L" pong"; }
};

struct RemotePong : Pingable {
  wstring ping(const wstring& message) override {
    wstring result;
    /* Microsoft REST SDK
            http_client client(U("https//localhsot:9149/"))
            uri_builder builder(U("/api/pingpong/"));
            pplx::task<wstring> task = client.request(
                    methods::GET, builder.to_string())
                    .then([=](http_response r){
                            return r.extract_string();
                    });
            task.wait();
            return task.get();

    */
    return L"";
  }
};

void tryit(Pingable& pp) { wcout << pp.ping(L"ping") << endl; }

void main(void) {
  Pong pp;
  for (int i = 0; i < 3; ++i) {
    tryit(pp);
  }
}
};  // namespace CommunicationProxy

int main(void) {
  Proxy::main();
  AttributeProxy::main();
  CachedProxy::main();
  VirtualProxy::main();
  CommunicationProxy::main();

  return 0;
}